### PR TITLE
First draft of pjit

### PR DESCRIPTION
**Pull request by skye**
Note that this is split up into two commits, one for `pjit` and one for `with_sharding_constraint` (which includes the `pjit` translation rule for testing). It may be easier to review them one at a time.

Overall this works and was super easy thanks to Adam's xmap work! I'd still like to get some more thoughts on the API, i.e. how the positional axis -> mesh axis name mapping should look and what it should be called.

### CODE REVIEW

File: jax/experimental/pjit.py [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+               for out, mesh_axes in safe_zip(out_nodes, out_mesh_axes)]
+
+  subc = subc.build(xops.Tuple(subc, out_nodes))
+  return xops.Call(c, subc, list(in_nodes))
+
+pjit_call_p = core.CallPrimitive("pjit_call")
+pjit_call_p.def_impl(_pjit_call_impl)
+xla.call_translations[pjit_call_p] = _pjit_translation_rule
+
+# None indicates replicated axis (partial replication)
```
Comment (apaszke) [inline jax/experimental/pjit.py:139]: I don't think that this is the same thing as partial replication. You have to have at least one `None` in your axis partitioning to even have a chance of triggering partial replication, but there are partitions that do use Nones and are sharded over the whole mesh. What matters is whether you use all mesh axis names or not!

File: jax/experimental/pjit.py [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+from ..lib import xla_bridge as xb
+from ..lib import xla_client as xc
+from ..tree_util import tree_flatten, tree_unflatten
+from .._src.util import (extend_name_stack, HashableFunction, safe_zip,
+                         wrap_name, wraps)
+xops = xc._xla.ops
+
+# TODO(skye): better name than mesh_axes
+def pjit(fun: Callable,
+         in_mesh_axes,
```
Comment (apaszke) [inline jax/experimental/pjit.py:37]: Those are supposed to be like `xmap`'s `in_axes`, but the leaves are `PartitionSpec`s of mesh axis names and `None`s, right?

File: jax/experimental/pjit.py [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+  if mesh_axes is None:
+    return OrderedDict()
+  return OrderedDict((axis, i)
+                     for i, axis_or_axes in enumerate(mesh_axes)
+                     for (axis, i) in _array_mapping_entries(axis_or_axes, i))
+
+def _array_mapping_entries(partitioning: ArrayAxisPartitioning, i: int):
+  if partitioning is None:
+    return
+  # TODO(skye): do axis names have to be strings?
```
Comment (apaszke) [inline jax/experimental/pjit.py:197]: They don't. They can be anything and we only recurse when they're a `list` or a `tuple`.

File: jax/experimental/pjit.py [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+sharding_constraint_p = core.Primitive("sharding_constraint")
+sharding_constraint_p.def_impl(_sharding_constraint_impl)
+sharding_constraint_p.def_abstract_eval(lambda x, **unused: x)
+ad.deflinear2(sharding_constraint_p,
+              lambda ct, _, mesh_axes, resource_env: (
+                  with_sharding_constraint(ct, mesh_axes),))
+xla.translations[sharding_constraint_p] = _sharding_constraint_translation_rule
+
+
+def get_array_mapping(mesh_axes: ArrayPartitioning) -> pxla.ArrayMapping:
```
Comment (apaszke) [inline jax/experimental/pjit.py:187]: This gets called all over the place. Why can't we just use the same API as in xmap and avoid converting back and forth between partition specs and the dicts?

File: jax/experimental/pjit.py [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+  in_avals = [core.raise_to_shaped(core.get_aval(arg)) for arg in args]
+  return _pjit_callable(
+      fun, in_mesh_axes, out_mesh_axes_thunk, resource_env, name,
+      *in_avals)(*args)
+
+def _pjit_translation_rule(c, axis_env, in_nodes, name_stack, backend, name,
+                           call_jaxpr, in_mesh_axes, out_mesh_axes_thunk,
+                           resource_env):
+  # TODO(skye): this should arguably be part of axis_env
+  mesh = resource_env.physical_mesh
```
Comment (apaszke) [inline jax/experimental/pjit.py:107]: I don't think it should. Adding those axes to `axis_env` would mean that the collective lowering rules could pick them up, but `pjit` doesn't introduce any axis names into scope!

**Review by apaszke — COMMENTED** [review]
Nice!

**Review by skye — COMMENTED** [review]
Forgot to publish these replies earlier :(

**Comment by skye** [comment]
@apaszke if you're happy with the pxla.py change, I propose we get this in as-is and iterate. I wanna continue iterating on the API and other changes but I think it'll be easier if we can experiment with checked-in code and review smaller chunks.

File: jax/experimental/pjit.py (hunk 5 above)
Comment (apaszke) [reply jax/experimental/pjit.py:107]: We currently do have an invariant that you can't switch the `resource_env` on any primitive wrapped in `xmap`/`pjit`, so I think it's fine. I'd just rather keep those mesh axes out of `axis_env`, because if we ever end up accidentally calling into e.g. a `psum` rule then it might succeed for no good reason instead of throwing an unbound axis error.

**Review by apaszke — APPROVED** [review]

File: jax/experimental/pjit.py (hunk 2 above)
Comment (jekbradbury) [reply jax/experimental/pjit.py:37]: I would maybe call these arguments `in_axis_resources` and `out_axis_resources` for max consistency with `xmap`?

**Review by jekbradbury — APPROVED** [review]

File: jax/experimental/pjit.py (hunk 2 above)
Comment (skye) [reply jax/experimental/pjit.py:37]: Yeah ok. I actually find the term `axis_resources` kinda confusing, but I'll use it here for consistency and maybe we can reconsider it for both `xmap` and `pjit` at some point.

[bikeshedding on naming below]
After studying the xmap docstring, I think the "axis" in `axis_resources` refers to the named array axes? Elsewhere in the docstring, the mesh axes are called "resource axes". So I think `axis_resources` is short for `axis_resource_axes`? If so, IMO renaming to `resource_axes` is slightly clearer.

File: jax/experimental/pjit.py [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+    out_mesh_axes = tuple(out_mesh_axes)
+
+  static_argnums = _ensure_index_tuple(static_argnums)
+
+  @wraps(fun)
+  def wrapped(*args, **kwargs):
+    if kwargs:
+      raise NotImplementedError("pjit over kwargs not yet supported")
+
+    # Putting this outside of fun_mapped would make resources lexically scoped
```
Comment (apaszke) [inline jax/experimental/pjit.py:61]: nit: there's no `fun_mapped` here

Comment (skye) [reply jax/experimental/pjit.py]: Oops, changed to `wrapped`

File: jax/experimental/pjit.py (hunk 4 above)
Comment (skye) [reply jax/experimental/pjit.py]: After talking to Adam, we concluded it should be pretty easy to use the xmap API and get the best of all worlds. I'm gonna do this in a follow-up to get this in while I figure out how xmap does it.

File: tests/pjit_test.py [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+    x = jnp.arange(np.prod(x_shape)).reshape(x_shape)
+    y = jnp.arange(np.prod(y_shape)).reshape(y_shape)
+    actual = f(x, y)
+    expected = x @ y
+    self.assertAllClose(actual, expected, check_dtypes=False)
+    self.assertIsInstance(actual, pxla.ShardedDeviceArray)
+    self.assertLen(actual.device_buffers, 4)
+
+    split0, split1 = np.split(expected, 2)
+    self.assertAllClose(actual.device_buffers[0].to_py(), split0,
```
Comment (gnecula) [inline tests/pjit_test.py:92]: I am wondering if it is specified that the device_buffers will be assigned as tested here?

File: tests/pjit_test.py [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+    self.assertIsInstance(actual, pxla.ShardedDeviceArray)
+    self.assertLen(actual.device_buffers, 2)
+    self.assertAllClose(actual.device_buffers[0].to_py(), expected,
+                        check_dtypes=False)
+
+  @with_mesh([('x', 2), ('y', 2)])
+  def testBasic2D(self):
+    @partial(pjit,
+             in_axis_resources=(P(None, 'x', 'y'), P('y')),
+             out_axis_resources=P('x'))
```
Comment (gnecula) [inline tests/pjit_test.py:77]: Is this an example of the partial replication feature, where the output is sharded on the x resource axis and replicated on the y?

File: jax/experimental/pjit.py [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+  return xb.set_sharding_proto(c, x_node,
+                               get_sharding_proto(c, x_node, axis_resources, mesh))
+
+sharding_constraint_p = core.Primitive("sharding_constraint")
+sharding_constraint_p.def_impl(_sharding_constraint_impl)
+sharding_constraint_p.def_abstract_eval(lambda x, **unused: x)
+ad.deflinear2(sharding_constraint_p,
+              lambda ct, _, axis_resources, resource_env: (
+                  with_sharding_constraint(ct, axis_resources),))
+xla.translations[sharding_constraint_p] = _sharding_constraint_translation_rule
```
Comment (gnecula) [inline jax/experimental/pjit.py:186]: Does it make sense to add this only to the "tpu" translation rules? Or, throw an error from the translation rule for non-TPU

File: jax/experimental/pjit.py [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+
+
+def with_sharding_constraint(x, axis_resources):
+  resource_env = maps.thread_resources.env
+  return sharding_constraint_p.bind(x, axis_resources=axis_resources,
+                                    resource_env=resource_env)
+
+def _sharding_constraint_impl(x, axis_resources, resource_env):
+  # TODO(skye): can we also prevent this from being called in other
+  # non-pjit contexts? (e.g. pmap, control flow)
```
Comment (gnecula) [inline jax/experimental/pjit.py:171]: Is pjit going to work eventually inside pmap, or xmap?

File: tests/pjit_test.py (hunk 8 above)
Comment (skye) [reply tests/pjit_test.py:77]: Yah, so easy with the mesh API :)

File: tests/pjit_test.py (hunk 7 above)
Comment (skye) [reply tests/pjit_test.py:92]: I don't think this is part of the public API, no. A less implementation-specific way to do this might be to look at each device_buffer's device to determine which split it should have, although that might make the test hard to reason about. I will probably focus on other things first before revisiting this test, but something to keep in mind if/when we do revisit it (e.g. if it breaks due to this detail :wink: ).

File: jax/experimental/pjit.py (hunk 10 above)
Comment (skye) [reply jax/experimental/pjit.py:171]: Good question! I thought so, but after our discussion with James and Anselm, maybe pjit-of-vmap is enough. We should think more about this.

File: jax/experimental/pjit.py (hunk 9 above)
Comment (skye) [reply jax/experimental/pjit.py:186]: Agh, I realize I forgot to copy the TPU check from `sharded_jit`'s impl to `pjit`'s impl. I think a top-level check will be better than a TPU-only translation rule, since this rule will theoretically work if/when XLA supports GPU sharding. I'll add  the impl check next time I touch this code.
